### PR TITLE
First approach to create a modulr payment

### PR DESCRIPTION
**Pull request by Franx0**
[DP-354](https://devengers.atlassian.net/browse/DP-354)

Transfer service creates payments in Modulr API. I decided to maintain transfer name in order to be consistent with other internal gems and becasue in our case we are creating associated transfer and not payment, for now.

#### How to test it ####
```ruby
require "modulr"

client=Modulr::Client.new({
  base_url: "https://api-sandbox.modulrfinance.com/api-sandbox-token",
  apikey: ENV["MODULR_SANDBOX_APIKEY"]
})

client.transfers.create(
  account_id: ENV["MODULR_ACCOUNT_ID"],
  currency: "EUR",
  amount: 1,
  destination_type: "IBAN",
  destination_iban: "GB20MODR00000000000001",
  destination_name: "Test",
  reference: "Salary"
)
```

### CODE REVIEW

File: spec/modulr/api/transfers_service_spec.rb [hunk 1]
```diff
@@ -0,0 +1,70 @@
+# frozen_string_literal: true
+
+RSpec.describe Modulr::API::TransfersService, :unit, type: :client do
+  subject(:transfers) { described_class.new(client) }
+
+  let(:client) { instance_double(Modulr::Client) }
+  let(:response) { Struct.new(body: , keyword_init: true) }
+  let(:modulr_response) { response.new(body: { data: fixture_response }) }
+  let(:options) { {} }
+  let(:fixture_response) do
```
Comment (Franx0) [inline spec/modulr/api/transfers_service_spec.rb:10]: Do we want a fixture system to check modulr API responses?

File: spec/modulr/api/transfers_service_spec.rb [hunk 2]
```diff
@@ -0,0 +1,70 @@
+# frozen_string_literal: true
+
+RSpec.describe Modulr::API::TransfersService, :unit, type: :client do
+  subject(:transfers) { described_class.new(client) }
+
+  let(:client) { instance_double(Modulr::Client) }
+  let(:response) { Struct.new(body: , keyword_init: true) }
+  let(:modulr_response) { response.new(body: { data: fixture_response }) }
```
Comment (Franx0) [inline spec/modulr/api/transfers_service_spec.rb:8]: Maybe, it would usefull to create a fixture helper to handle this 👍🏻

File: lib/modulr.rb [hunk 3]
```diff
@@ -3,17 +3,6 @@
 require_relative "modulr/version"
 
 require_relative "modulr/auth/signature"
-
```
Comment (Franx0) [inline lib/modulr.rb:6]: I cleaned a little bit how we required different files in modulr dir

Comment (TechiRexach) [reply lib/modulr.rb:6]: None will be used more than once than the required ones that you have taken from the file?

File: spec/modulr/api/transfers_service_spec.rb (hunk 1 above)
Comment (Nachooc5) [reply spec/modulr/api/transfers_service_spec.rb:10]: The mocking system that we have in Devengo Client seems cleaner to me https://github.com/devengoapp/devengo_client/blob/81cc3384a473e9fdf88fd362b540dca59b12ca50/spec/lib/core/services/authentication_service_spec.rb#L20

**Review by Nachooc5 — APPROVED** [review]

File: lib/modulr/api/transfers_service.rb [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+module Modulr
+  module API
+    class TransfersService < Service
+      def create( # rubocop:disable Metrics/ParameterLists
+        account_id:,
+        currency:,
+        amount:,
+        destination_type:,
+        destination_iban:,
+        destination_name:,
```
Comment (Franx0) [inline lib/modulr/api/transfers_service.rb:12]: destination fields should be received as a VO? In that case I find better to receive base fields instead a composed object, what do you think @devengoapp/dev ?

File: spec/modulr/api/transfers_service_spec.rb (hunk 1 above)
Comment (aitor) [reply spec/modulr/api/transfers_service_spec.rb:10]: I mentioned in Slack that I want all these client gems to evolve to use `webmock` and HTTP files cf. https://devengo.slack.com/archives/CHLUNJ0KS/p1658003395168969?thread_ts=1658003275.097419&cid=CHLUNJ0KS

File: lib/modulr.rb (hunk 3 above)
Comment (aitor) [reply lib/modulr.rb:6]: For gems, it's pretty common to centralize the requirements in the main lib file, but I'm ok with dispersing if the rest of the team finds it more precise and more maintainable.

File: lib/modulr/api/transfers_service.rb (hunk 4 above)
Comment (aitor) [reply lib/modulr/api/transfers_service.rb:12]: As an API consumer, I prefer to deal with plain hashes:

```rb
Whatever::API.create({
  one: 'XYZ',
  another: 'HEY',
  destination: {
    just_another: { and_another: true },
    but_more: { and_more: true },
  },
})
```
It's good and noble we enforce as much strictness as we want in our final requests to the APIs, but I don't want to transfer that strictness and complexity to the consumer and the creation of its client code.

File: lib/modulr/resources/transfers/transfer.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+module Modulr
+  module Resources
+    module Transfers
+      class Transfer < Base
+        attr_accessor :id, :externalReference, :status # rubocop:disable Naming/MethodName
+
+        alias payment_reference_id id
+        alias external_reference externalReference
```
Comment (aitor) [inline lib/modulr/resources/transfers/transfer.rb:10]: Do NOT use the attr_accessor/alias pattern. Please replace it with the map helper:

```suggestion
        map :id, [:id, :payment_reference_id]
        map :status, :status
        map :externalReference, :external_reference
```

File: spec/fixtures/transfers/201_transfer_create.json [hunk 6]
```diff
@@ -0,0 +1,15 @@
+{
```
Comment (aitor) [inline spec/fixtures/transfers/201_transfer_create.json:1]: As mentioned in the PR, we should progressively move to use HTTP files and webmock.

**Review by aitor — CHANGES_REQUESTED** [review]

File: lib/modulr/api/transfers_service.rb (hunk 4 above)
Comment (Franx0) [reply lib/modulr/api/transfers_service.rb]: Done

File: lib/modulr/client.rb [hunk 7]
```diff
@@ -70,6 +73,9 @@ def request_options(_method, _path, data, _options)
     end
 
     def add_auth_options!(options)
+      options[:headers][:authorization] = @apikey
```
Comment (Franx0) [inline lib/modulr/client.rb:76]: If requested url is sandbox modulr environment we do not need to send signature headers but authorization with apikey value instead.

File: spec/modulr/api/transfers_service_spec.rb (hunk 1 above)
Comment (Franx0) [reply spec/modulr/api/transfers_service_spec.rb:10]: I think you mean the same we maintain in devengo_client gem, where we mock all real request with webmock and return response through fixtures. If it is the case I'll do in another PR due to the changes we need to add to implement it 👍🏻

File: lib/modulr/api/services.rb [hunk 8]
```diff
@@ -1,11 +1,19 @@
...
+require_relative "transfers_service"
+
 module Modulr
   module API
     module Services
       def accounts
         @services[:accounts] ||= API::AccountsService.new(self)
       end
+
+      def transfers
```
Comment (IvanGuardado) [inline lib/modulr/api/services.rb:14]: IMO, we should not be consistent between API clients. Clients should represent the much they can the provider API.

Comment (Franx0) [reply lib/modulr/api/services.rb:14]: ok, payments instead is the choosen name 👍🏻

**Comment by IvanGuardado** [comment]
I'm merging this in order to start including the gem in the project. We can iterate on the given feedback later.

**Pull request merged**